core/src/com/unciv/ui/worldscreen/mainmenu/DropBox.kt

```diff
@@ -151,3 +173,112 @@ class OnlineMultiplayer {
+/**
+ * Used to communicate data access between players
+ */
+class LockFile {
+    // The lockData is necessary to make every LockFile unique
+    // If Dropbox gets a file with the same content and overwrite set to false, it returns no
+    // error even though the file was not uploaded as the exact file is already existing
+    var lockData = UUID.randomUUID().toString()
+}
+
+/**
+ *	Wrapper around OnlineMultiplayer's synchronization facilities.
+ *
+ *	Based on the design of Mutex from kotlinx.coroutines.sync, except that when it blocks,
+ *	  it blocks the entire thread for an increasing period of time via Thread.sleep()
+ */
+class ServerMutex(val gameInfo: GameInfoPreview) {
+    private var locked = false
+
+    constructor(gameInfo: GameInfo) : this(gameInfo.asPreview()) { }
+
+    /**
+     * Try to obtain the server lock ONCE
+     * DO NOT forget to unlock it when you're done with it!
+     * Sleep between successive attempts
+     * @see lock
+     * @see unlock
+     * @return true if lock is acquired
+     */
+    fun tryLock(): Boolean {
+        // If we already hold the lock, return without doing anything
+        if (locked) {
+            return locked
+        }
+        
+        locked = false
+        val fileName = "${OnlineMultiplayer().getGameLocation(gameInfo.gameId)}_Lock"
+
+        // We have to check if the lock file already exists before we try to upload a new
+        // lock file to not overuse the dropbox file upload limit else it will return an error
+        if (DropBox.fileExists(fileName)) {
+            return locked
+        }
+
+        try {
+            DropBox.uploadFile(fileName, Gzip.zip(GameSaver.json().toJson(LockFile())))
+        } catch (ex: DropBoxFileConflictException) {
+            return locked
+        }
+
+        locked = true
+        return locked
+    }
+
+    /**
+     * Block until this client owns the lock
+     *
+     * TODO: Create an alternative to the underlying tryLock or tryLockGame which checks for
+     *       (and returns, when present) the value of the Retry-After header
+     *
+     * @see tryLock
+     * @see unlock
+     */
+    fun lock() {
+        var tries = 0
+
+        // Try the lockfile once
+        locked = tryLock()
+        while (!locked) {
+            // Wait exponentially longer after each attempt as per DropBox API recommendations
+            var delay = 250 * 2.0.pow(tries).toLong()
+
+            // 8 seconds is a really long time to sleep a thread, it's as good a cap as any
+            if (delay > 8000) {
+                delay = 8000
+            }
+
+            // Consider NOT sleeping here, instead perhaps delay or spin+yield
+            Thread.sleep(delay)
+
+            tries++
+
+            // Retry the lock
+            locked = tryLock()
+        }
+    }
+
+    /**
+     * Release a server lock acquired by tryLock or lock
+     * @see tryLock
+     * @see lock
+     */
+    fun unlock() {
+        if (!locked)
+            return
```

Comment:
Personal preference: I'd throw an exception to crash early here if any users of this lock tried to unlock it without having acquired it.

…Hm… This might be horrible practice, so maybe don't do this, but I think throwing an exception in a new thread would get us Google Play logs without crashing the client?

Reply:
I don't think calling unlock when not previously locked is a problem we really need to care about. 
The mutex should be run in a separate thread anyway as lock() blocks it, which would result in ANRs and that thread probably also has a catch inside it as it's mandatory when working with the OnlineMultiplayer class as basically everything throws exceptions, so I don't think we would get Google Play logs. Or would we? When do Google Play logs get generated?

Reply:
> […] so I don't think we would get Google Play logs. Or would we? When do Google Play logs get generated?

I haven't confirmed this personally, but it seems that log entries are generated whenever any thread (including the main thread) exits due to an uncaught exception, which also means that you can force a log entry by rethrowing an exception in a short-lived thread (`catch (e: Throwable) { thread { throw e } }`).

https://github.com/yairm210/Unciv/pull/5833#issuecomment-1001064624

https://support.google.com/googleplay/android-developer/answer/9859174